packages/compiler/core/checker.ts

```diff
@@ -1789,6 +1960,1115 @@ export function createChecker(program: Program): Checker {
+        type.nodeByKind.set("Union", node);
+        break;
+      case SyntaxKind.ProjectionInterfaceSelector:
+        projectionsByTypeKind.get("Interface")!.push(node);
+        type.nodeByKind.set("Interface", node);
+        break;
+      default:
+        const projected = checkTypeReference(node.selector);
+        let current = projectionsByType.get(projected);
+        if (!current) {
+          current = [];
+          projectionsByType.set(projected, current);
+        }
+        current.push(node);
+        type.nodeByType.set(projected, node);
+        break;
+    }
+
+    return links.declaredType;
+  }
+
+  function evalProjectionNode(
+    node: ProjectionExpression | ProjectionStatementItem
+  ): TypeOrReturnRecord {
+    switch (node.kind) {
+      case SyntaxKind.ProjectionExpressionStatement:
+        return evalProjectionExpressionStatement(node);
+      case SyntaxKind.ProjectionCallExpression:
+        return evalProjectionCallExpression(node);
+      case SyntaxKind.ProjectionMemberExpression:
+        return evalProjectionMemberExpression(node);
+      case SyntaxKind.Identifier:
+        return evalProjectionIdentifier(node);
+      case SyntaxKind.ProjectionLambdaExpression:
+        return evalProjectionLambdaExpression(node);
+      case SyntaxKind.StringLiteral:
+        return evalStringLiteral(node);
+      case SyntaxKind.NumericLiteral:
+        return evalNumericLiteral(node);
+      case SyntaxKind.ProjectionBlockExpression:
+        return evalProjectionBlockExpression(node);
+      case SyntaxKind.ProjectionArithmeticExpression:
+        return evalProjectionArithmeticExpression(node);
+      case SyntaxKind.ProjectionIfExpression:
+        return evalProjectionIfExpression(node);
+      case SyntaxKind.ProjectionEqualityExpression:
+        return evalProjectionEqualityExpression(node);
+      case SyntaxKind.ProjectionUnaryExpression:
+        return evalProjectionUnaryExpression(node);
+      case SyntaxKind.ProjectionRelationalExpression:
+        return evalProjectionRelationalExpression(node);
+      case SyntaxKind.ProjectionReference:
+        return evalProjectionReferenceExpression(node);
+      case SyntaxKind.VoidKeyword:
+        return voidType;
+      case SyntaxKind.NeverKeyword:
+        return neverType;
+      case SyntaxKind.Return:
+        return evalReturnKeyword(node);
+      default:
+        throw new Error("Can't eval " + SyntaxKind[node.kind]);
+    }
+  }
+
+  interface EvalContext {
+    node: Node;
+    locals: Map<string, Type>;
+    parent?: EvalContext;
+  }
+
+  function evalReturnKeyword(node: ReturnExpressionNode): ReturnRecord {
+    const value = evalProjectionNode(node.value);
+    if (value.kind === "Return") {
+      return value;
+    }
+
+    return {
+      kind: "Return",
+      value,
+    };
+  }
+
+  function evalProjectionReferenceExpression(
+    node: ProjectionReferenceExpressionNode
+  ): TypeOrReturnRecord {
```

Comment:
We have `TypeOrReturnRecord` to let the TS compiler ensure we handle return records properly, so they never bubble out to users.